S3 POST policy

riquito opened the pull request:
Add the ability to generate a [AWS POST Policy](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTConstructPolicy.html). 

AWS POST Policy allows to define some stringent rules like limiting the size of S3 uploads (see issue #1503)

Here I'm proposing a builder approach, not dissimilar by the approach taken by MinIO. 

Since no services are required and the nature of the job, I think that rusoto_signature is a good candidate to host it.

I introduced Chrono as dependency because by other discussions it looked like the intention is to migrate to it and I didn't want to add technical debt with a new feature.

Usage example:

```
use rusoto_signature::PostPolicy;
use rusoto_signature::Region;
use chrono::{TimeZone, Utc};

let expiration_date = Utc.ymd(2020, 1, 1).and_hms(1, 2, 3);
let res = PostPolicy::default()
    .set_bucket_name("some-bucket-name")
    .set_region(&Region::EuCentral1)
    .set_access_key_id("aws-access-key")
    .set_secret_access_key("aws-secret-key")
    .set_key("object-key")
    .set_expiration(&expiration_date)
    .set_content_length_range(0, 1024 * 1024)
    .build_form_data();
let (upload_url, form_data) = res.unwrap()
```

TODO
- [ ] Create the POST URL properly, reusing existing code
- [x] Add tests
- [x] Add docs
- [x] Update CHANGELOG

riquito commented:
I rebased to match master and re-run the CI run on github actions (semaphoreci is failing on other PRs too).

How does the builder approach look for this functionality? Any comment?

riquito commented on rusoto/signature/Cargo.toml:
```diff
@@ -33,19 +33,20 @@ log = "0.4.1"
 pin-project = "0.4"
 percent-encoding = "2"
 tokio = { version = "0.2", features = ["macros"] }
+chrono = "0.4"
```
Perhaps it would be better to follow the example of https://github.com/rusoto/rusoto/pull/1829 and remove default features

phyber replied:
Yeah, if the optional features aren't needed, it's probably nicer to disable them.

zeroclock commented:
Amazing feature, @riquito !

May I suggest to add `session token` ( `x-amz-security-token` ) to PostPolicy struct?
It's required when using a policy generated with temporary credentials in Lambda environment etc.

https://docs.aws.amazon.com/STS/latest/APIReference/CommonParameters.html

Sample implementation (It works fine under my Lambda environment):
https://github.com/zeroclock/rusoto/commit/3ccfb90cc8f2bb94867a869784b2687092327e9b

riquito commented:
Thanks @zeroclock, nice addition! Since in some cases (Amazon DevPay) ` x-amz-security-credential` can accept multiple tokens comma separated, I changed its name to "security_token" and refactored your implementation to reuse an utility function.

shekohex commented:
Any Updates on this PR? is it still alive?

shekohex commented:
I ended up with something crazy since I need this feature for an app I'm building, I made an `ext` module for `rusoto_s3` and ended up copying this Code to the ext, and it works very well!

This, of course, a temporary solution until this PR gets merged and a new version gets released.

Again, Thank you @riquito for your Work!

riquito commented:
Rebased, fixed conflicts from master. Now using just chrono for dates.

Thanks @shekohex for the feedback! The PR is alive, I was just waiting for a maintainer to step up (unfortunately rusoto entered maintenance mode instead). Here hoping someone will give support to this project

greglearns commented:
@riquito and @shekohex given that rusoto is in maintenance mode, and the new AWS library doesn't really support S3 yet, would you recommend forking rusoto or using the `ext` module approach (@shekohex can you explain that approach more)?

shekohex commented:
Hello @greglearns 
To do so, I have an `ext` module, where I put any `TraitExt` or other simple functionality to extended a crate that I use.
In this case, you could simply create a `rusoto_s3` submodule of the `ext` module and add the code from this PR.

Here a slightly modified version of the module that I use in my application.

```rust
//! AWS Post Policy
//!
//! Follows [POST Policy
//! format](http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTConstructPolicy.html)
//! using AWS Signature 4 algorithm.

#![allow(unused)]

use chrono::{DateTime, NaiveDate, Utc};
use hmac::{Hmac, Mac, NewMac};
use rusoto_core::region::Region;
use rusoto_signature::SignedRequest;
use serde::ser::{SerializeSeq, Serializer};
use serde::Serialize;
use sha2::Sha256;
use std::collections::HashMap;

#[derive(Default)]
pub struct PostPolicy<'a> {
    expiration: Option<&'a DateTime<Utc>>,
    content_length_range: Option<(u64, u64)>,
    conditions: Vec<Condition<'a>>,
    form_data: HashMap<String, String>,
    bucket_name: Option<&'a str>,
    key: Option<&'a str>,
    region: Option<&'a Region>,
    access_key_id: Option<&'a str>,
    secret_access_key: Option<&'a str>,
}

#[derive(Serialize)]
pub struct SerializablePolicy<'a> {
    expiration: &'a str,
    conditions: &'a Vec<Condition<'a>>,
}

struct Condition<'a>((&'a str, &'a str, &'a str));

impl<'a> Serialize for Condition<'a> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut seq = serializer.serialize_seq(Some(3))?;
        let v = &self.0;
        seq.serialize_element(v.0)?;

        if v.0 == "content-length-range" {
            seq.serialize_element(&v.1.parse::<u64>().map_err(|_| {
                serde::ser::Error::custom(
                    "expected u64 value, the minimum content length",
                )
            })?)?;
            seq.serialize_element(&v.2.parse::<u64>().map_err(|_| {
                serde::ser::Error::custom(
                    "expected u64 value, the maximum content length",
                )
            })?)?;
        } else {
            seq.serialize_element(v.1)?;
            seq.serialize_element(v.2)?;
        }

        seq.end()
    }
}

/// ```
///     // Create a POST policy to limit uploads to up to 1MiB
///     # use rusoto_signature::PostPolicy;
///     # use rusoto_signature::Region;
///     # use chrono::{TimeZone, Utc};
///
///     let expiration_date = Utc.ymd(2020, 1, 1).and_hms(1, 2, 3);
///
///     let res = PostPolicy::default()
///         .set_bucket_name("some-bucket-name")
///         .set_region(&Region::EuCentral1)
///         .set_access_key_id("aws-access-key")
///         .set_secret_access_key("aws-secret-key")
///         .set_key("object-key")
///         .set_expiration(&expiration_date)
///         .set_content_length_range(0, 1024 * 1024)
///         .build_form_data();
///
///     let (upload_url, form_data) = res.unwrap();
/// ```
impl<'a> PostPolicy<'a> {
    /// Set expiration time policy condition
    pub fn set_expiration(mut self, t: &'a DateTime<Utc>) -> Self {
        self.expiration = Some(t);
        self
    }

    /// Set key policy condition and update the form data
    pub fn set_key(mut self, key: &'a str) -> Self {
        if key.is_empty() {
            return self;
        }

        self = self.append_policy("eq", "$key", &key);
        self.key = Some(key);
        self.form_data.insert("key".to_string(), key.to_string());
        self
    }

    /// Set key starts-with policy condition and update the form data
    #[allow(dead_code)]
    pub fn set_key_startswith(mut self, key: &'a str) -> Self {
        if key.is_empty() {
            return self;
        }

        self.key = Some(key);

        self = self.append_policy("starts-with", "$key", &key);
        self.form_data.insert("key".to_string(), key.to_string());
        self
    }

    /// Set bucket name policy condition and update the form data
    pub fn set_bucket_name(mut self, bucket_name: &'a str) -> Self {
        self.form_data
            .insert("bucket".to_string(), bucket_name.to_string());
        self = self.append_policy("eq", "$bucket", bucket_name);
        self.bucket_name = Some(bucket_name);
        self
    }

    /// Set region
    pub fn set_region(mut self, region: &'a Region) -> Self {
        self.region = Some(region);
        self
    }

    /// Set access key id
    pub fn set_access_key_id(mut self, access_key_id: &'a str) -> Self {
        if access_key_id.is_empty() {
            return self;
        }

        self.access_key_id = Some(access_key_id);
        self
    }

    /// Set secret access key
    pub fn set_secret_access_key(mut self, secret_access_key: &'a str) -> Self {
        if secret_access_key.is_empty() {
            return self;
        }

        self.secret_access_key = Some(secret_access_key);
        self
    }

    /// Set security token
    pub fn set_security_token(mut self, security_token: &'a str) -> Self {
        if security_token.is_empty() {
            return self;
        }

        self.form_data.insert(
            "x-amz-security-token".to_string(),
            security_token.to_string(),
        );
        self =
            self.append_policy("eq", "$x-amz-security-token", security_token);
        self
    }

    /// Set content-type policy condition and update the form data
    #[allow(dead_code)]
    pub fn set_content_type(mut self, content_type: &'a str) -> Self {
        self.form_data
            .insert("Content-Type".to_string(), content_type.to_string());
        self = self.append_policy("eq", "$Content-Type", content_type);
        self
    }

    /// Set content length range policy condition and update the form data
    pub fn set_content_length_range(
        mut self,
        min_length: u64,
        max_length: u64,
    ) -> Self {
        self.content_length_range = Some((min_length, max_length));
        // We should append the policy here, but ownership is tricky,
        // so we'll do it inside build_form_data()
        self
    }

    /// Append policy condition
    ///
    /// ```
    /// # use rusoto_signature::PostPolicy;
    ///
    /// PostPolicy::default()
    ///     .append_policy("eq", "$acl", "public-read");
    /// ```
    ///
    /// Depending on the kind of condition you may have to add
    /// the field to the form data produced by
    /// [build_form_data](#method.build_form_data)
    pub fn append_policy(
        mut self,
        match_type: &'a str,
        target: &'a str,
        value: &'a str,
    ) -> Self {
        self.conditions.push(Condition((match_type, target, value)));
        self
    }

    /// Create the form data using the policy
    pub fn build_form_data(
        mut self,
    ) -> Result<(String, HashMap<String, String>), String> {
        match self.content_length_range {
            Some((min_length, max_length)) if min_length > max_length => {
                return Err(format!(
                    "Min-length ({}) must be <= Max-length ({})",
                    min_length, max_length
                ));
            },
            _ => (),
        }

        if self.expiration.is_none() {
            return Err("Expiration date must be specified".to_string());
        }

        if self.key.is_none() {
            return Err("Object key must be specified".to_string());
        }

        if self.bucket_name.is_none() {
            return Err("Bucket name must be specified".to_string());
        }

        if self.region.is_none() {
            return Err("Region must be specified".to_string());
        }

        if self.access_key_id.is_none() {
            return Err("Access key id must be specified".to_string());
        }

        if self.secret_access_key.is_none() {
            return Err("Secret access key must be specified".to_string());
        }

        let bucket_name = self.bucket_name.unwrap();
        let secret_access_key = self.secret_access_key.unwrap();

        let expiration = self
            .expiration
            .unwrap()
            .format("%Y-%m-%dT%H:%M:%S.000Z")
            .to_string();

        let current_time = if cfg!(test) {
            use chrono::TimeZone;
            Utc.ymd(2020, 1, 1).and_hms(0, 0, 0)
        } else {
            Utc::now()
        };
        let current_time_fmted =
            current_time.format("%Y%m%dT%H%M%SZ").to_string();
        let current_date = current_time.format("%Y%m%d").to_string();

        let access_key_id = self.access_key_id.unwrap();
        let region = self.region.unwrap();
        let region_name = region.name();

        let x_amz_credential = format!(
            "{}/{}/{}/{}/aws4_request",
            &access_key_id, &current_date, &region_name, "s3",
        );

        let mut conditions: Vec<Condition> =
            self.conditions.into_iter().collect();

        conditions.push(Condition(("eq", "$x-amz-date", &current_time_fmted)));
        conditions.push(Condition((
            "eq",
            "$x-amz-algorithm",
            "AWS4-HMAC-SHA256",
        )));
        conditions.push(Condition((
            "eq",
            "$x-amz-credential",
            &x_amz_credential,
        )));

        let min_length_as_string: String;
        let max_length_as_string: String;
        if let Some((min, max)) = self.content_length_range {
            min_length_as_string = min.to_string();
            max_length_as_string = max.to_string();
            conditions.push(Condition((
                "content-length-range",
                &min_length_as_string,
                &max_length_as_string,
            )))
        }

        let policy_to_serialize = SerializablePolicy {
            expiration: &expiration,
            conditions: &conditions,
        };

        let policy_as_json = serde_json::to_string(&policy_to_serialize)
            .map_err(|e| format!("{:?}", e))?;

        let policy_as_base64 = base64::encode(policy_as_json);

        let signature_date = current_time.date().naive_utc();

        let x_amz_signature = sign_string(
            &policy_as_base64,
            &secret_access_key,
            signature_date,
            &region_name,
            "s3",
        );

        self.form_data
            .insert("policy".to_string(), policy_as_base64);
        self.form_data
            .insert("x-amz-date".to_string(), current_time_fmted);
        self.form_data.insert(
            "x-amz-algorithm".to_string(),
            "AWS4-HMAC-SHA256".to_string(),
        );
        self.form_data
            .insert("x-amz-credential".to_string(), x_amz_credential);
        self.form_data
            .insert("x-amz-signature".to_string(), x_amz_signature);

        let signed_request = SignedRequest::new("GET", "s3", &region, "/");
        let upload_url = if cfg!(feature = "aws-local") {
            // use path format for the url.
            // since we are running on localhost anyway!
            format!(
                "{}://{}/{}",
                signed_request.scheme(),
                signed_request.hostname(),
                bucket_name
            )
        } else {
            format!(
                "{}://{}.{}",
                signed_request.scheme(),
                bucket_name,
                signed_request.hostname()
            )
        };

        Ok((upload_url, self.form_data))
    }
}

#[inline]
fn hmac(secret: &[u8], message: &[u8]) -> Hmac<Sha256> {
    let mut hmac =
        Hmac::<Sha256>::new_varkey(secret).expect("failed to create hmac");
    hmac.update(message);
    hmac
}

/// Takes a message and signs it using AWS secret, time, region keys and service
/// keys.
fn sign_string(
    string_to_sign: &str,
    secret: &str,
    date: NaiveDate,
    region: &str,
    service: &str,
) -> String {
    let date_str = date.format("%Y%m%d");
    let date_hmac = hmac(
        format!("AWS4{}", secret).as_bytes(),
        date_str.to_string().as_bytes(),
    )
    .finalize()
    .into_bytes();
    let region_hmac = hmac(date_hmac.as_ref(), region.as_bytes())
        .finalize()
        .into_bytes();
    let service_hmac = hmac(region_hmac.as_ref(), service.as_bytes())
        .finalize()
        .into_bytes();
    let signing_hmac = hmac(service_hmac.as_ref(), b"aws4_request")
        .finalize()
        .into_bytes();
    hex::encode(
        hmac(signing_hmac.as_ref(), string_to_sign.as_bytes())
            .finalize()
            .into_bytes(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::prelude::*;

    const BUCKET: &str = "the-bucket";
    const REGION: Region = Region::EuCentral1;
    const ACCESS_KEY_ID: &str = "foo_access_key";
    const SECRET_ACCESS_KEY: &str = "foo_secret_key";
    const OBJECT_KEY: &str = "the-object-key";

    #[test]
    fn bucket_name_is_required() {
        let expiration_date = Utc.ymd(2020, 1, 1).and_hms(1, 2, 3);

        let res = PostPolicy::default()
            .set_region(&REGION)
            .set_access_key_id(ACCESS_KEY_ID)
            .set_secret_access_key(SECRET_ACCESS_KEY)
            .set_key(OBJECT_KEY)
            .set_expiration(&expiration_date)
            .build_form_data();

        assert_eq!(res, Err("Bucket name must be specified".to_string()));
    }

    #[test]
    fn region_is_required() {
        let expiration_date = Utc.ymd(2020, 1, 1).and_hms(1, 2, 3);

        let res = PostPolicy::default()
            .set_bucket_name(&BUCKET)
            .set_access_key_id(ACCESS_KEY_ID)
            .set_secret_access_key(SECRET_ACCESS_KEY)
            .set_key(OBJECT_KEY)
            .set_expiration(&expiration_date)
            .build_form_data();

        assert_eq!(res, Err("Region must be specified".to_string()));
    }
    #[test]
    fn access_key_id_is_required() {
        let expiration_date = Utc.ymd(2020, 1, 1).and_hms(1, 2, 3);

        let res = PostPolicy::default()
            .set_bucket_name(&BUCKET)
            .set_region(&REGION)
            .set_secret_access_key(SECRET_ACCESS_KEY)
            .set_key(OBJECT_KEY)
            .set_expiration(&expiration_date)
            .build_form_data();

        assert_eq!(res, Err("Access key id must be specified".to_string()));
    }

    #[test]
    fn secret_access_key_is_required() {
        let expiration_date = Utc.ymd(2020, 1, 1).and_hms(1, 2, 3);

        let res = PostPolicy::default()
            .set_bucket_name(&BUCKET)
            .set_region(&REGION)
            .set_access_key_id(ACCESS_KEY_ID)
            .set_key(OBJECT_KEY)
            .set_expiration(&expiration_date)
            .build_form_data();

        assert_eq!(res, Err("Secret access key must be specified".to_string()));
    }

    #[test]
    fn expiration_is_required() {
        let res = PostPolicy::default()
            .set_bucket_name(&BUCKET)
            .set_region(&REGION)
            .set_access_key_id(ACCESS_KEY_ID)
            .set_key(OBJECT_KEY)
            .build_form_data();

        assert_eq!(res, Err("Expiration date must be specified".to_string()));
    }
    #[test]
    fn build_successfully() {
        let expiration_date = Utc.ymd(2020, 1, 1).and_hms(1, 2, 3);

        let res = PostPolicy::default()
            .set_bucket_name(BUCKET)
            .set_region(&REGION)
            .set_access_key_id(ACCESS_KEY_ID)
            .set_secret_access_key(SECRET_ACCESS_KEY)
            .set_key(OBJECT_KEY)
            .set_expiration(&expiration_date)
            .set_content_length_range(123, 456)
            .build_form_data();

        assert!(res.is_ok());
        let (upload_url, form_data) = res.unwrap();
        if cfg!(feature = "aws-local") {
            assert_eq!(
                upload_url,
                "https://s3.eu-central-1.amazonaws.com/the-bucket"
            );
        } else {
            assert_eq!(
                upload_url,
                "https://the-bucket.s3.eu-central-1.amazonaws.com"
            );
        }
        assert_eq!(form_data.get("key").unwrap(), "the-object-key");

        assert_eq!(form_data.get("bucket").unwrap(), "the-bucket");
        assert_eq!(
            form_data.get("x-amz-algorithm").unwrap(),
            "AWS4-HMAC-SHA256"
        );
        assert_eq!(
            form_data.get("x-amz-credential").unwrap(),
            "foo_access_key/20200101/eu-central-1/s3/aws4_request"
        );
        assert_eq!(form_data.get("x-amz-date").unwrap(), "20200101T000000Z");

        let expected_policy = serde_json::json!({
            "expiration": "2020-01-01T01:02:03.000Z",
            "conditions": [
                ["eq", "$bucket", "the-bucket"],
                ["eq", "$key", "the-object-key"],
                ["eq", "$x-amz-date", "20200101T000000Z"],
                ["eq", "$x-amz-algorithm", "AWS4-HMAC-SHA256"],
                ["eq", "$x-amz-credential", "foo_access_key/20200101/eu-central-1/s3/aws4_request"],
                ["content-length-range", 123, 456]
            ]
        });

        let policy_as_base64 = form_data.get("policy").unwrap();
        let policy_as_vec_u8 = base64::decode(policy_as_base64).unwrap();
        let policy: serde_json::Value =
            serde_json::from_slice(&policy_as_vec_u8).unwrap();
        assert_eq!(policy, expected_policy);
    }

    #[test]
    fn set_content_type() {
        let expiration_date = Utc.ymd(2020, 1, 1).and_hms(1, 2, 3);

        let res = PostPolicy::default()
            .set_content_type("some/type")
            .set_bucket_name(BUCKET)
            .set_region(&REGION)
            .set_access_key_id(ACCESS_KEY_ID)
            .set_secret_access_key(SECRET_ACCESS_KEY)
            .set_key(OBJECT_KEY)
            .set_expiration(&expiration_date)
            .build_form_data();

        assert!(res.is_ok());

        let (_, form_data) = res.unwrap();
        dbg!(&form_data);
        assert_eq!(form_data.get("Content-Type").unwrap(), "some/type");

        let policy_as_base64 = form_data.get("policy").unwrap();
        let policy_as_vec_u8 = base64::decode(policy_as_base64).unwrap();
        let policy: serde_json::Value =
            serde_json::from_slice(&policy_as_vec_u8).unwrap();
        let conditions = policy["conditions"].as_array().unwrap();
        assert!(conditions.contains(&serde_json::json!([
            "eq",
            "$Content-Type",
            "some/type"
        ])));
    }

    #[test]
    fn append_policy() {
        let expiration_date = Utc.ymd(2020, 1, 1).and_hms(1, 2, 3);

        let res = PostPolicy::default()
            .append_policy("a", "b", "c")
            .set_bucket_name(BUCKET)
            .set_region(&REGION)
            .set_access_key_id(ACCESS_KEY_ID)
            .set_secret_access_key(SECRET_ACCESS_KEY)
            .set_key(OBJECT_KEY)
            .set_expiration(&expiration_date)
            .build_form_data();

        let (_, form_data) = res.unwrap();

        assert_eq!(form_data.get("a"), None);

        let policy_as_base64 = form_data.get("policy").unwrap();
        let policy_as_vec_u8 = base64::decode(policy_as_base64).unwrap();
        let policy: serde_json::Value =
            serde_json::from_slice(&policy_as_vec_u8).unwrap();
        let conditions = policy["conditions"].as_array().unwrap();
        assert!(conditions.contains(&serde_json::json!(["a", "b", "c"])));
    }

    #[test]
    fn set_key_startswith() {
        let expiration_date = Utc.ymd(2020, 1, 1).and_hms(1, 2, 3);

        let res = PostPolicy::default()
            .set_key_startswith("foo")
            .set_bucket_name(BUCKET)
            .set_region(&REGION)
            .set_access_key_id(ACCESS_KEY_ID)
            .set_secret_access_key(SECRET_ACCESS_KEY)
            .set_expiration(&expiration_date)
            .build_form_data();

        let (_, form_data) = res.unwrap();
        dbg!(&form_data);
        assert_eq!(form_data.get("key").unwrap(), "foo");

        let policy_as_base64 = form_data.get("policy").unwrap();
        let policy_as_vec_u8 = base64::decode(policy_as_base64).unwrap();
        let policy: serde_json::Value =
            serde_json::from_slice(&policy_as_vec_u8).unwrap();
        let conditions = policy["conditions"].as_array().unwrap();
        assert!(conditions.contains(&serde_json::json!([
            "starts-with",
            "$key",
            "foo"
        ])));
    }

    #[test]
    fn set_security_token() {
        let expiration_date = Utc.ymd(2020, 1, 1).and_hms(1, 2, 3);
        let security_token = "some-session-token";

        let res = PostPolicy::default()
            .set_bucket_name(BUCKET)
            .set_region(&REGION)
            .set_access_key_id(ACCESS_KEY_ID)
            .set_secret_access_key(SECRET_ACCESS_KEY)
            .set_key(OBJECT_KEY)
            .set_expiration(&expiration_date)
            .set_security_token(&security_token)
            .build_form_data();

        let (_, form_data) = res.unwrap();
        dbg!(&form_data);
        assert_eq!(
            form_data.get("x-amz-security-token").unwrap(),
            "some-session-token"
        );

        let policy_as_base64 = form_data.get("policy").unwrap();
        let policy_as_vec_u8 = base64::decode(policy_as_base64).unwrap();
        let policy: serde_json::Value =
            serde_json::from_slice(&policy_as_vec_u8).unwrap();
        let conditions = policy["conditions"].as_array().unwrap();
        assert!(conditions.contains(&serde_json::json!([
            "eq",
            "$x-amz-security-token",
            "some-session-token"
        ])));
    }
}
```

and here is a snippet where I use it

```rust
...
use crate::ext::rusoto_s3::PostPolicy;
...
let random_key = nanoid::nanoid();
let key = format!("{}.{}", random_key, ext);
let bucket = media_type.to_string();
let region = current.aws_region();
let credentials = current.aws_credentials_provider()?.credentials().await?;
let expiration_date = Utc::now() + Duration::minutes(10);
let content_length = data.size_in_bytes as u64;
let (url, form_data) = PostPolicy::default()
    .set_content_type(content_type)
    .set_bucket_name(&bucket)
    .set_region(&region)
    .set_access_key_id(credentials.aws_access_key_id())
    .set_secret_access_key(credentials.aws_secret_access_key())
    .set_key(&key)
    .set_expiration(&expiration_date)
    .set_content_length_range(content_length, content_length + 1)
    .build_form_data()
    .map_err(Error::Rusoto)?;
// now send the URL and form_data to the client.

...
```

Hope it helps you too!

greglearns commented:
@shekohex Thank you so much, this is perfect!

c1wren commented:
Can this please get merged? This is extremely useful and works really well.
